framework/include/utils/MooseTypes.h

```diff
@@ -64,6 +65,39 @@
                beginIndex_2(__VA_ARGS__),                                                          \
                beginIndex_1(__VA_ARGS__),                                                          \
                beginIndex_0(__VA_ARGS__))
+
+/**
+ * MooseIndex Macro for creating an index type of the right type for loops and other places where
+ * type matching is important.
+ * Usage:
+ *
+ * Type t;
+ *
+ * Container type:
+ * for (MooseIndex(t) i = 0; i < t.size(); ++i)
+ *
+ * POD type:
+ * for (MooseIndex(t) i = 0; i < t; ++i)
+ */
+#define MooseIndex(type) decltype(_MooseIndex(type, 0))
+
+// SFINAE templates for type MooseIndex type selection
+template <typename T, typename std::enable_if<std::is_integral<T>::value>::type * = nullptr>
+typename std::remove_const<T>::type
```

Comment:
We might consider also wrapping with `std::remove_reference` although it should be rare that we're ever calling this macro on an integral reference...